### ISSUE TITLE
[MCOMPILER-41] recompiling behaves differently when build.sourceDirectory is not set to the default path src/main/java/

### ISSUE THREAD

**Issue by olamy** [olamy/maven-compiler-plugin#210]
**[akinokaze](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=akinkaze)** opened **[MCOMPILER-41](https://issues.apache.org/jira/browse/MCOMPILER-41?redirect=false)** and commented

the default build.sourceDirectory is at src/man/java
When our code is located in this default directory, the first time we run mvn compile all the source codes will be compiled as expected.
When we run mvn compile again without any changes to the src code, the message
"Nothing to compile - all classes are up to date", which is also the expected behaviour.

If we set build.sourceDirectory to another directory, the behaviour is different.
The first time we run mvn compile all the source codes will be compiled as expected.
But when we run mvn compile again without any changes to the src code, all the source codes are compiled again.
This will increase the rebuild time significantly.

---

**Attachments:**
- [MavenWebTest.zip](https://issues.apache.org/jira/secure/attachment/12714891/MavenWebTest.zip) (_4.08 kB_)
- [MCOMPILER-41.zip](https://issues.apache.org/jira/secure/attachment/12714867/MCOMPILER-41.zip) (_658 bytes_)

**Comment by olamy** [comment]
**[Benjamin Bentmann](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=bentmann)** commented

Not reproducable with the attached mini project (Maven Compiler Plugin 2.0.2).

**Comment by olamy** [comment]
**[Olivier Lamy](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=olamy)** commented

We can't reproduce it.
If you can provide a test project to reproduce, please reopen the issue.

**Comment by olamy** [comment]
**[Kelly Davis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kldavis4)** commented

I am attaching a new test (MavenWebTest) which demonstrates the issue.

I am running version 2.0.7 in MacOSX 10.5.6 with Jdk 1.5.

Run 'mvn compile' multiple times and it will recompile the unchanged source file every time. If you comment out the sourceDirectory property (and move the java directory to src/main) it only compiles it the first time. Please test and reopen this bug.

**Comment by olamy** [comment]
**[Kelly Davis](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kldavis4)** commented

I did some playing with the source for the plexus-compiler-api to see what was happening. Basically the issue appears to be a result of specifying multiple source roots. If I specify a sourceDirectory as src and in src I actually have 2 (or more) source directories, everything compiles fine, but the mapping between source and target files are screwed up (because I really have multiple source directories). So when the scanner checks to see if the file(s) need to be compiled, they aren't there. I fixed the sourceDirectory to specify just one of my source directories under src, and then use the buildhelper plugin to add the additional source directories, and it is now detecting changes properly. The only potential issue here, I guess, is that it would be nice if it would detect and give a warning when you specify multiple source directories.
